### PR TITLE
Pass SecurityContext parameter to AccessControl

### PR DESCRIPTION
**Pull request by kokosing**
Pass Session parameter to AccessControl

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/Session.java [hunk 1]
```diff
@@ -293,6 +293,31 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
         requireNonNull(transactionManager, "transactionManager is null");
         requireNonNull(accessControl, "accessControl is null");
 
+        Session sessionWithTransaction = new Session(
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/Session.java]: return it also instead of creating new session object

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/Session.java]: revert

File: presto-main/src/main/java/io/prestosql/Session.java [hunk 2]
```diff
@@ -315,7 +340,7 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
 
             for (Entry<String, String> property : catalogProperties.entrySet()) {
                 // verify permissions
-                accessControl.checkCanSetCatalogSessionProperty(transactionId, identity, catalogName, property.getKey());
+                accessControl.checkCanSetCatalogSessionProperty(sessionWithTransaction, catalogName, property.getKey());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/Session.java]: revert. instead pass `identity` to `checkCanSetCatalogSessionProperty`

File: presto-main/src/main/java/io/prestosql/Session.java [hunk 3]
```diff
@@ -331,7 +356,7 @@ public Session beginTransactionId(TransactionId transactionId, TransactionManage
                     .orElseThrow(() -> new PrestoException(NOT_FOUND, "Catalog does not exist: " + catalogName))
                     .getConnectorId();
             if (role.getType() == SelectedRole.Type.ROLE) {
-                accessControl.checkCanSetRole(transactionId, identity, role.getRole().get(), catalogName);
+                accessControl.checkCanSetRole(sessionWithTransaction, role.getRole().get(), catalogName);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/Session.java]: revert. instead pass `identity` to `checkCanSetRole `

**Comment by dain** [comment]
Can you describe the use case for this?   We have discussed this in the past a few times and always decided to not expose session to access control.

File: presto-main/src/main/java/io/prestosql/Session.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/Session.java]: `checkCanSetCatalogSessionProperty` requires admin level privileges (write access on entire catalog), it would be a pity to not store that access in audit storage.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/Session.java]: Passing a session that is not fully formed yet looks error-prone to me. Let's discuss.

**Comment by kokosing** [comment]
> Can you describe the use case for this? We have discussed this in the past a few times and always decided to not expose session to access control.

There are access control implementations that are logging information about access checks to audit service. That way it is possible to audit what user accessed what database resources. Along storing information about resource (table, column etc), action (select, insert) and user (identity), it is also useful to log what was the actual query that was executed, the client ip etc.

I see that passing Session might be not a best option because: 
 - it passes to much information to access control
 - leads to a chicken-egg-problem, see https://github.com/prestosql/presto/pull/171#discussion_r254231730 (session is needed to create session o_O)

Maybe instead of passing Session, we could pass some SessionContext with information required for audining but not leaking information like session properties etc.

@electrum @dain @findepi @sopel39 Thoughts?

**Comment by kokosing** [comment]
> Maybe instead of passing Session, we could pass some SessionContext with information required for audining but not leaking information like session properties etc.

SecurityContext is a better name.

File: presto-main/src/main/java/io/prestosql/Session.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/Session.java]: Agree. See https://github.com/prestosql/presto/pull/171#issuecomment-461359314

**Comment by dain** [comment]
> > Can you describe the use case for this? We have discussed this in the past a few times and always decided to not expose session to access control.
> 
> There are access control implementations that are logging information about access checks to audit service. That way it is possible to audit what user accessed what database resources. Along storing information about resource (table, column etc), action (select, insert) and user (identity), it is also useful to log what was the actual query that was executed, the client ip etc.

Interesting use case.  We have always used the QueryCompletionEvent to log this information.  Maybe @martint or @electrum remember exactly how you extract the read and written tables (partitions?) in the events.

> I see that passing Session might be not a best option because:
> * it passes to much information to access control
> * leads to a chicken-egg-problem, see [#171 (comment)](https://github.com/prestosql/presto/pull/171#discussion_r254231730) (session is needed to create session o_O)

Do you have a list of information that is necessary for this?

**Comment by findepi** [comment]
> Interesting use case. We have always used the QueryCompletionEvent to log this information. Maybe @martint or @electrum remember exactly how you extract the read and written tables (partitions?) in the events.

@dain for security audit purposes, you want to log _before_ query execution.

**Comment by kokosing** [comment]
> Do you have a list of information that is necessary for this?

Not much today, however I assume the list may grow once it will be more widely used. Today, besides currently available information (resource, user, access type) I use sql query and client ip.

**Comment by dain** [comment]
> > Interesting use case. We have always used the QueryCompletionEvent to log this information. Maybe @martint or @electrum remember exactly how you extract the read and written tables (partitions?) in the events.
> 
> @dain for security audit purposes, you want to log _before_ query execution.

Interesting.  How will you know if the user actually accessed the data?  For example, if write a query over table A, which I have permissions, and B, which I don't, you will get an access control check for A, which will work , and B, which will fail, but in the end _no_ data was accessed because the query failed during analysis.

Anyway, I believe you are going with the `SecurityContext` approach above.  If so, let me know when it is ready, and I'll take a look.

**Comment by kokosing** [comment]
@dain @electrum I changed this to pass `SecuritySession` (new entity) to `ConnectorAccessControl`. Currently `SecuritySession` so far wraps only `ConnectorIdentity`. 

PS. This PR needs some more polishing, I will take care of it once you approve the direction.

**Comment by kokosing** [comment]
@dain, @electrum ping

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorAccessControl.java [hunk 4]
```diff
@@ -53,280 +52,280 @@
 public interface ConnectorAccessControl
 {
     /**
-     * Check if identity is allowed to create the specified schema in this catalog.
+     * Check if session is allowed to create the specified schema in this catalog.
```
Comment (electrum) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorAccessControl.java]: Saying *session* here feels wrong, maybe it because it implies we are making a decision on something based other than identity. I'd rather keep the original word *identity*.

File: presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedAccessControl.java [hunk 5]
```diff
@@ -31,11 +31,10 @@
 
 public class TestFileBasedAccessControl
 {
-    public static final ConnectorTransactionHandle TRANSACTION_HANDLE = new ConnectorTransactionHandle() {};
+    private static final ConnectorTransactionHandle TRANSACTION_HANDLE = new ConnectorTransactionHandle() {};
```
Comment (electrum) [inline presto-plugin-toolkit/src/test/java/io/prestosql/plugin/base/security/TestFileBasedAccessControl.java]: These cleanups are unrelated, can you move to a separate commit?

File: presto-main/src/main/java/io/prestosql/Session.java (hunk 3 above)
Comment (electrum) [reply presto-main/src/main/java/io/prestosql/Session.java]: I agree, constructing the session twice seems like a hack.

**Comment by electrum** [comment]
Thanks for updating this. The overall direction looks good, but I still have concerns.

We should also update `SystemAccessControl` so that the interfaces are consistent.

I like the originally proposed name `SecurityContext` better, although I'm struggling to articulate why. It would give us a place to put additional information needed in the future, even if that information doesn't seem to be part of a "session".

Related, we could make `ConnectorTransactionId` a part of that interface to simplify the argument list. Though we'd need `ConnectorSecurityContext` and `SecurityContext`. But that might be a good idea if there will be additional connector-specific information in the future (that wouldn't be appropriate for `SystemAccessControl`).

Another problem I have is passing `Session` into all the internal `AccessControl` methods. It seems bad to widen the interfaces to such as huge object as `Session`. Creating the session twice for `checkCanSetCatalogSessionProperty()` is one example of a problem this causes.

**Comment by dain** [comment]
> We should also update `SystemAccessControl` so that the interfaces are consistent.

Agreed, but can you do that as a second commit.... that first one is already pretty long
 
> I like the originally proposed name `SecurityContext` better, although I'm struggling to articulate why. It would give us a place to put additional information needed in the future, even if that information doesn't seem to be part of a "session".

I agree.  The information isn't really a "session" but a feels to me like additional context information to make security decission.  The naming also aligns with `ConnectorContext` passed to `ConnectorFactory#create`

> Related, we could make `ConnectorTransactionId` a part of that interface to simplify the argument list. Though we'd need `ConnectorSecurityContext` and `SecurityContext`. But that might be a good idea if there will be additional connector-specific information in the future (that wouldn't be appropriate for `SystemAccessControl`).

I think we should split the interface regardless.  I'm not confident about `ConnectorTransactionId`, but it seems it seems better than not putting it in the context.

**Comment by dain** [comment]
please reassign when this is ready to be reviewed again.

**Comment by kokosing** [comment]
> We should also update SystemAccessControl so that the interfaces are consistent.

Let me do it once we are all good with `ConnectorAccessControl` and `AccessControl`. I do not want to do it too fast to avoid commit conflicts solving in case we change our mind.

> I like the originally proposed name SecurityContext better

Sounds good to me. 👍 

> Related, we could make ConnectorTransactionId a part of that interface to simplify the argument list. Though we'd need ConnectorSecurityContext and SecurityContext.

That was a good idea. Code of `AccessControlManager` got simplified a lot.

> Another problem I have is passing Session into all the internal AccessControl methods.

To solve that we would need to have yet another `SecurityContext` class. Do you suggest to have `ConnectorSecurityContext`, `SystemSecurityContext` and `SecurityContext`?

**Comment by kokosing** [comment]
@electrum and @dain I updated the pull request. Please take a look.

> Agreed, but can you do that as a second commit.

I would even go further with that and handle that as separate pull request. Maintainance of such big refactorings is very time consuming.

**Comment by findepi** [comment]
@kokosing please update the commit message and PR title (one refers to `SecuritySession`, the other to `Session`; but neither is being used)

**Comment by kokosing** [comment]
@dain @electrum ping

**Comment by dain** [comment]
I'm confused about this change now.  It appears to be simply refactoring the following:
* Change `AccessControl` and the main implementation `AccessControlManager` to take `Session` instead of `TransactionId` and `Identity`.
* Change `ConnectorSecurityContext` to take `ConnectorSecurityContext` instead of `ConnectorTransactionHandle` and `ConnectorIdentity`.
* No change to `SystemAccessControl` which still takes just `Identity`.

I'm confused because I thought we needed new information.

Anyway, does the above correctly summarize the point of this change?  Are there plans to add a `SystemSecurityContext`?

**Comment by kokosing** [comment]
> I'm confused because I thought we needed new information.

You are right. This is only a refactor, that should allow us easily add (or remove) information that are passed to connector and system access control.

> does the above correctly summarize the point of this change?

Yes. 

* `AccessControl` takes `Session`. But my plan is to pass `SecurityContext` there. I have made that change because I wanted to make sure that you agree with this.
* `ConnectorAccessControl` takes `ConnectorSecurityContext`.
* `SystemAccessControl` will take `SystemSecurityContext`. Not implemented yet, also I would like to hear your opinion first.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java [hunk 6]
```diff
@@ -67,198 +66,197 @@ public LegacyAccessControl(
...
+        Optional<Table> target = metastoreProvider.apply(((HiveTransactionHandle) securityContext.getTransactionHandle())).getTable(tableName.getSchemaName(), tableName.getTableName());
 
         if (!target.isPresent()) {
             denyDropTable(tableName.toString(), "Table not found");
         }
 
-        if (!identity.getUser().equals(target.get().getOwner())) {
-            denyDropTable(tableName.toString(), "Owner of the table is different from session user");
+        if (!securityContext.getIdentity().getUser().equals(target.get().getOwner())) {
+            denyDropTable(tableName.toString(), "Owner of the table is different from securityContext user");
```
[the remark this answers is not in the document]
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java]: good catch! It was changed by search-and-replace. I need to make sure there are no other things like that.

**Comment by dain** [comment]
This seems fine to me.  I don't have any further comments.

**Comment by kokosing** [comment]
> This seems fine to me. I don't have any further comments.

@dain Thanks!

**Comment by kokosing** [comment]
@dain @electrum @findepi Implemented all the pending comments, please review.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java [hunk 7]
```diff
@@ -67,198 +66,197 @@ public LegacyAccessControl(
...
 
-        Optional<Table> target = metastoreProvider.apply(((HiveTransactionHandle) transaction)).getTable(tableName.getSchemaName(), tableName.getTableName());
+        Optional<Table> target = metastoreProvider.apply(((HiveTransactionHandle) securityContext.getTransactionHandle())).getTable(tableName.getSchemaName(), tableName.getTableName());
 
         if (!target.isPresent()) {
             denyDropTable(tableName.toString(), "Table not found");
         }
 
-        if (!identity.getUser().equals(target.get().getOwner())) {
+        if (!securityContext.getIdentity().getUser().equals(target.get().getOwner())) {
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/LegacyAccessControl.java]: I'd prefer to rename the parameter `securityContext` to just `context` in all of the interfaces and implementations. The shorter name would seem to make the code more readable (the name itself seems unambiguous).
